### DISCUSSION TITLE
'str' object has no attribute 'comments'

### DISCUSSION THREAD

**Discussion by HelloMukama** [Radi85/Comment#192] [Q&A] [answered]
So, i have added everything to this page https://django-comment-dab.readthedocs.io/Web%20API.html

bu i keep getting the error in title.
How can i resolve this?
Is it necessary to keep on with the whole customization thing which is on the following pages?
Or is it me doing it wrong?

i cloned this github page and pasted all the updated comment...html files in the comment folder you say we shd create from the docs.
Still the error is the same.
I installed all requirements but nothing seems to change
Why? Someone kindly enlighten me.

**Comment by Radi85** [comment]
Hey @hellomukama,

Are you using the api or the UI? Can you please share the traceback?

It seems to me that you didn't pass the right object to `render_comments` in the template.

> Is it necessary to keep on with the whole customization thing which is on the following pages?

No. This is only needed if you would like to customize the current UI.

**Reply by HelloMukama** [reply]
i did a pip install.
`/home/nmj/.local/lib/python3.9/site-packages/comment/templatetags/comment_tags.py, line 61, in get_comments_count
```
@register.simple_tag(name='get_comments_count')
def get_comments_count(obj, user):
    return obj.comments.all_comments_by_object(obj, include_flagged=is_comment_moderator(user)).count() …
```
`

The error seems to be rising from the comments attr in the obj object in the above return statement

Below is the traceback

```
Internal Server Error: /submissions/submission-19/details/
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.9/dist-packages/django/core/handlers/base.py", line 204, in _get_response
    response = response.render()
  File "/usr/local/lib/python3.9/dist-packages/django/template/response.py", line 105, in render
    self.content = self.rendered_content
  File "/usr/local/lib/python3.9/dist-packages/django/template/response.py", line 83, in rendered_content
    return template.render(context, self._request)
  File "/usr/local/lib/python3.9/dist-packages/django/template/backends/django.py", line 61, in render
    return self.template.render(context)
  File "/usr/local/lib/python3.9/dist-packages/django/template/base.py", line 170, in render
    return self._render(context)
  File "/usr/local/lib/python3.9/dist-packages/django/template/base.py", line 162, in _render
    return self.nodelist.render(context)
  File "/usr/local/lib/python3.9/dist-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/lib/python3.9/dist-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/lib/python3.9/dist-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/usr/local/lib/python3.9/dist-packages/django/template/base.py", line 162, in _render
    return self.nodelist.render(context)
  File "/usr/local/lib/python3.9/dist-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/lib/python3.9/dist-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/lib/python3.9/dist-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/usr/local/lib/python3.9/dist-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/lib/python3.9/dist-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/lib/python3.9/dist-packages/django/template/library.py", line 192, in render
    output = self.func(*resolved_args, **resolved_kwargs)
  File "/home/nmj/.local/lib/python3.9/site-packages/comment/templatetags/comment_tags.py", line 61, in get_comments_count
    return obj.comments.all_comments_by_object(obj, include_flagged=is_comment_moderator(user)).count()
AttributeError: 'str' object has no attribute 'comments'
```

I also tried cloning the whole thing onto my local dir and it seems to work except for the pagination.html not found error
But the one of the pip installation is just not working

**Reply by HelloMukama** [reply]
Now my qn is.
I have this simple system i am building and according to the docs of this app, everything was supposed to be pretty simple till i just failed to see anything.
QN:
Can i really integrate this app in my already existing model such that each time i load the model details.html page, i get the little comment box below? Thanks bro

**Reply by Radi85** [reply]
As I said above, you probably didn't pass the proper object to `render_comments` function. In order to render the comments in your template you need to use this template tag as described [here](https://django-comment-dab.readthedocs.io/usage.html#usage)
```jinja
{% render_comments object request %}
```
which `object` is the instance of the model that you use the comment with.

Can you show me how you used the `render_comments`?

> Can i really integrate this app in my already existing model such that each time i load the model details.html page, i get the little comment box below?

Yes. Our comment app can be associated with any django model.

**Reply by HelloMukama** [reply]
```
{% extends "base/base.html" %}
{% block title %} Submission | Details {% endblock %}
{% load static %}
{% load comment_tags %}  {# Loading the template tag #}
{% render_comments obj request %}  {# Render all the comments belonging to the passed object "obj" #}
{% block content %}
```

As the docs say, i simply added the above lines 4 & 5 at the top of my details.html page.
Then immediately after the submission description - 'description' is a field on my Submission model, i added `<h6>{% get_comments_count obj user %}</h6>`
And altogether, the error is `AttributeError: 'str' object has no attribute 'comments'` which for the get_comments_count tag, is understandable as it's calling the function that's triggering the error

**Reply by Radi85** [reply]
replace `obj` with the actual object in your details.html.
Or share the full `detail.html` here if you don't get what I mean.

**Reply by HelloMukama** [reply]
```
{% extends "base/base.html" %}
{% block title %} Submission | Details {% endblock %}
{% load static %}
{% load comment_tags %}  {# Loading the template tag #}
{% render_comments obj request %}  {# Render all the comments belonging to the passed object "obj" #}
{% block content %}

<div class="row ml-3 pl-lg-4">
    <div class="col-12 col-xl-12 ml-3">
        <div class="card card-body shadow-sm mb-4">
            <h6 class="mb-4">Details of this submission <span>as of {{time.now}}</span></h6>

            <div class="">
                <h6>About the User that made the submission</h6>
                <p>Full Names: {{ submission.profile.user.first_name }} {{ submission.profile.user.last_name }}</p>
                <p>Email Address: {% if not submission.profile.email %}
                    <span class="text-danger">User has no email address</span>{% else %}{{ submission.profile.email }}{% endif %}
                </p>
            </div>

            <div>
                <h6>The submission. <span style="font-size: small; color: darkblue;">Last edited on {{ submission.updated }}</span></h6>
                <p>Title: <span class="text-uppercase">{{ submission.submission }}</span></p>
                <p>Date Of Submission: <span style="color: darkblue;">{{ submission.timestamp }}</span></p>
            </div>

            <div>
                <h6>Details</h6>
                <p>{{ submission.description }}</p>
            </div>

            <div>
                <h6>{% get_comments_count obj user %}</h6>

            </div>
        </div>
    </div>
</div>

{% endblock %}

<!-- Specific Page JS goes HERE  -->
{% block javascripts %}{% endblock %}

```

**Reply by HelloMukama** [reply]
So i've changed obj to submission and it's showing a count of zero.
Next step is how to get the comment form into details.html

**Reply by HelloMukama** [reply]
tried looking through comment_tags.py and cant see any create_comment tag

**Reply by Radi85** [reply]
Updated the file below.

**Comment by Radi85** [comment, marked as answer]
@hellomukama,
Replace `obj` in the comment tags with `submission` your actual object.
Update your details.html as following:

Edit: Move the `{% render_comments submission request %}` inside the `{% block content %}` and get rid of `{% get_comments_count obj user %}`.
```jinja
{% extends "base/base.html" %}
{% block title %} Submission | Details {% endblock %}
{% load static %}
{% load comment_tags %}  {# Loading the template tag #}

{% block content %}

<div class="row ml-3 pl-lg-4">
    <div class="col-12 col-xl-12 ml-3">
        <div class="card card-body shadow-sm mb-4">
            <h6 class="mb-4">Details of this submission <span>as of {{time.now}}</span></h6>

            <div class="">
                <h6>About the User that made the submission</h6>
                <p>Full Names: {{ submission.profile.user.first_name }} {{ submission.profile.user.last_name }}</p>
                <p>Email Address: {% if not submission.profile.email %}
                    <span class="text-danger">User has no email address</span>{% else %}{{ submission.profile.email }}{% endif %}
                </p>
            </div>

            <div>
                <h6>The submission. <span style="font-size: small; color: darkblue;">Last edited on {{ submission.updated }}</span></h6>
                <p>Title: <span class="text-uppercase">{{ submission.submission }}</span></p>
                <p>Date Of Submission: <span style="color: darkblue;">{{ submission.timestamp }}</span></p>
            </div>

            <div>
                <h6>Details</h6>
                <p>{{ submission.description }}</p>
            </div>

            <div>
                {% render_comments submission request %}  {# Use the actual object#}
            </div>
        </div>
    </div>
</div>

{% endblock %}

<!-- Specific Page JS goes HERE  -->
{% block javascripts %}{% endblock %}
```

**Reply by HelloMukama** [reply]
It's still weird coz even in the admin when i try to create a new comment, i get all fields but none asks for the submission to which to attach the comment. Why is that?

**Reply by HelloMukama** [reply]
Perfect. I applied your edit and im now getting the error below

```
Error during template rendering
In template /home/nmj/.local/lib/python3.9/site-packages/comment/templates/comment/comments/create_comment.html, error at line 4

'comment' is not a registered namespace
```

**Reply by HelloMukama** [reply]
someone kindly get back to me on this issue. thanks

**Reply by abhiabhi94** [reply]
you probably have added `comment` urls in some other `urls.py` but not in your base `urls.py` file. Also, whenever you're facing an issue, consider posting the `traceback`, so that others don't just have to shoot in the dark.

**Reply by HelloMukama** [reply]
Error during template rendering
In template /home/nmj/.local/lib/python3.9/site-packages/comment/templates/comment/comments/create_comment.html, error at line 4

`'comment' is not a registered namespace`

```
{% load comment_tags %}
{% load i18n %}
{% if is_anonymous_allowed or user.is_authenticated %}
<form class="js-comment-form" method="post" data-url="**{% url 'comment:create' %}**">
{% csrf_token %}
<div class="row">
<div class="{% block text_area_wrapper_cls %}col-sm-9 col-md-10 px-2 m-2 m-sm-0{% endblock text_area_wrapper_cls %}">
{% include 'comment/comments/comment_form.html' with placeholder=placeholder %}
</div>
<input name="model_name" value="{% get_model_name model_object %}" hidden>
<input name="model_id" value="{{ model_object.id }}" hidden>
<input name="app_name" value="{% get_app_name model_object %}" hidden>
<input name="oauth" value="{{oauth}}" hidden>
{% if comment %}
```


The above is the traceback that ive been getting in my template following directions from @Radi85 
And no, i dont have any other comment urls. matter of fact ive tried even commenting the api url from the docs of this `Comment app` but still get the same error

**Reply by HelloMukama** [reply]
OH MAN, I THINK YOU GUYS NEED TO EXPLICITLY INCLUDE THIS IN THE DOCS AS WELL
AFTER @abhiabhi94 TALKING ABOUT URLS, I REMEMBERED THAT USUALLY THE ERROR I AM GETTING WE ALWAYS GET WHEN THE URLS FILE OF THE APP IN QUESTION AREN'T BEING DETECTED.
SO , DID THIS IN MY MAIN URLS FILE 
`path('comment/', include('comment.urls')),`

AND BOOM I GOT THE COMMENT BOX NEXT TO MY MODEL'S DETAILS.
